module/VuFind/src/VuFind/Controller/MyResearchController.php

```diff
@@ -1527,6 +1605,42 @@ public function verifyAction()
         return $this->forwardTo('MyResearch', 'Login');
     }
 
+    /**
+     * Receive a hash and display the new password form if it's valid
+     *
+     * @return view
+     */
+    public function verifyEmailAction()
+    {
+        // If we have a submitted form
+        if ($hash = $this->params()->fromQuery('hash')) {
+            $hashtime = $this->getHashAge($hash);
+            $config = $this->getConfig();
+            // Check if hash is expired
+            $hashLifetime = isset($config->Authentication->recover_hash_lifetime)
+                ? $config->Authentication->recover_hash_lifetime
+                : 1209600; // Two weeks
+            if (time() - $hashtime > $hashLifetime) {
+                $this->flashMessenger()
+                    ->addMessage('recovery_expired_hash', 'error');
+                return $this->forwardTo('MyResearch', 'Login');
```

Comment:
Is there a strong reason that you chose `forwardTo` here, rather than something like `return $this->redirect()->toRoute('myresearch-home');`?

I ask because I ran into a scenario where I clicked a verification link and then was forced to log in twice before I could actually access the site. I think this may have been caused by some sub-optimal post-login redirect workflow that looped me back to the verification page after login even though I was already verified. Forcing a redirect to a different page instead of loading the login screen inside the currently executing action might work better.

Let me know if I'm being unclear or if you are unable to reproduce the state I'm describing. I can try some more detailed experiments on this end when I have a little more time, if necessary.

Reply:
I did not run into the described scenario, but i also had no particular reason to use forwardTo, so if you see the mentioned redirect method as more stable than it might be the better option.

Reply:
Each approach has advantages and disadvantages, but redirecting makes the relationship between urls and their contents more clear, so I think it's at least worth trying.